bigquery/google/cloud/bigquery/job.py

```diff
@@ -572,12 +588,20 @@ def schema(self, value):
         :type value: list of :class:`SchemaField`
         :param value: fields describing the schema
 
-        :raises: TypeError if 'value' is not a sequence, or ValueError if
-                 any item in the sequence is not a SchemaField
+        :raises TypeError: If ``value`is not a sequence.
+        :raises ValueError: If any item in the sequence is not
+                            a ``SchemaField``.
         """
-        if not all(isinstance(field, SchemaField) for field in value):
-            raise ValueError('Schema items must be fields')
-        self._schema = tuple(value)
+        if not value:
+            self._schema = ()
+        else:
+            if not all(isinstance(field, SchemaField) for field in value):
+                raise ValueError('Schema items must be fields')
+            if self._configuration is not None and self.autodetect:
```

Comment:
Thanks for the quick response.

> so to not create restrictions on the order the variables are created

IMO, the correct approach here is a "dumb" constructor that just sets values. We are "breaking" this approach by using the fancy `schema` `@property` for validation in the constructor. All methods on a class should be able to freely assume that the attributes set in the constructor will be present.

> it's easier to maintain a code with less restrictions

IMO it's easier to maintain code completely devoid of `getattr` / `setattr`

> Leaving `self._configuration` be loaded before schema makes its validation unnecessary

OK I'll send an update to this PR that removes that check